AudioManagerUtilities/src/CAmSocketHandler.cpp

```diff
@@ -114,80 +110,71 @@ CAmSocketHandler::~CAmSocketHandler()
 void CAmSocketHandler::start_listenting()
 {
     mDispatchDone = false;
-    int16_t pollStatus;
 
 #ifndef WITH_TIMERFD 
     clock_gettime(CLOCK_MONOTONIC, &mStartTime);
 #endif    
     timespec buffertime;
     
-    VectorListPoll_t cloneListPoll;
-    VectorListPoll_t::iterator listmPollIt;
-    VectorListPollfd_t fdPollingArray; //!<the polling array for ppoll
-    
-    auto preparePollfd = [&](const sh_poll_s& row)
-    {
-        CAmSocketHandler::prepare((sh_poll_s&)row);
-        pollfd temp = row.pollfdValue;
-        temp.revents = 0;
-        fdPollingArray.push_back(temp);
-    };
+    VectorPollfd_t fdPollingArray; //!<the polling array for ppoll
 
     while (!mDispatchDone)
     {
-        if (mRecreatePollfds)
+        // iterate all times though list and update ppoll accordingly
+        auto fdPollIt = fdPollingArray.begin();
+        for (auto it = mMapShPoll.begin(); it != mMapShPoll.end(); )
         {
-            fdPollingArray.clear();
-            //freeze mListPoll by copying it - otherwise we get problems when we want to manipulate it during the next lines
-            cloneListPoll = mListPoll;
-            //there was a change in the setup, so we need to recreate the fdarray from the list
-            std::for_each(cloneListPoll.begin(), cloneListPoll.end(), preparePollfd);
-            mRecreatePollfds = false;
-        }
-        else
-        {
-            //first we go through the registered filedescriptors and check if someone needs preparation:
-            std::for_each(cloneListPoll.begin(), cloneListPoll.end(), CAmSocketHandler::prepare);
+            auto& elem = it->second;
+            switch (elem.state)
+            {
+                case poll_states_e::REMOVE:
+                    close(elem.pollfdValue.fd);
+                    fdPollIt = fdPollingArray.erase(fdPollIt);
+                    it = mMapShPoll.erase(it);
+                    continue;
+
+                case poll_states_e::UNINIT:
+                    fdPollIt = fdPollingArray.emplace(fdPollIt);
+                    // fallthrough
+
+                case poll_states_e::UPDATE:
+                    CAmSocketHandler::prepare(elem);
+                    elem.state = poll_states_e::VALID;
+                    *fdPollIt = elem.pollfdValue;
+                    break;
+
+                default:
+                    break;
+            }
+            ++fdPollIt;
```

Comment:
What assumptions are made here...?  How do you know that you won't reach the end of fdPollingArray before the end of mMapShPoll?
--> Request changes (comment at least)

Reply:
This is not an assumption. The data will be similar all the time. Furthermore, if the vector is at its end and you increment the iterator will still point to end. No problem for the loop.
The only problem is once you try to delete the elements out of the fdPollingArray. This is caught in REMOVE case.
What would be now your preferred comment here?

Reply:
> This is not an assumption

Well, we may use language differently.  You might say you "know" (that data is similar).  I used the word "assumption" for the same thing - I'm not trying to imply anything.   The _code_ is written such that it requires an "assumed" relationship between the two data structures.  

But if there is no relationship, no problem (based then on that going ++ on end() is apparently safe)

> if the vector is at its end and you increment the iterator will still point to end

I could have sworn I've read that it's undefined behavior.  Is it true only for vectors?  

>What would be now your preferred comment here?

Not sure now...  If ++ on end() really means it stays on end() then I'd put that comment in perhaps.

If there was a relationship here (but you said there isn't), but if there were I might try to put this relationship into code somehow.  If data structures are related to each other, some of the program could show that by grouping them -  something like (not actual names of course):

```
struct pollstuff
{
   /* NOTE: These two data structures should mirror each other, i.e. always have the same size */ 
...or...
  /* NOTE If a handle exist in data struct X, it will always exist in Y */
    Map pollmap;
    Vector pollvector;
}
```

Another thing is accessing them only through a few abstraction functions, that are especially careful about the relationship (and asserting it).  That way no one makes the mistake of violating the relationship.   These are just generic ideas here and might not apply in this case.

If you're right about it being safe to go "beyond" end() then it's not an issue, but a comment could help.

Reply:
>> if the vector is at its end and you increment the iterator will still point to end
> I could have sworn I've read that it's undefined behavior. Is it true only for vectors?
You are right. Actually, an interrator can not know the content of the container.

> If you're right about it being safe to go "beyond" end() then it's not an issue, but a comment could help.
I have implemented a new test to ensure this - and found two issues in original implementation.

Reply:
Added more comments to explain the behavior. Also I have added an assert before the iterator with respective comment.